core/dispatcher/src/main/scala/whisk/core/dispatcher/ActivationFeed.scala

```diff
@@ -0,0 +1,100 @@
+ * limitations under the License.
+ */
+
+package whisk.core.dispatcher
+
+import scala.concurrent.duration.FiniteDuration
+import scala.language.postfixOps
+import scala.util.Try
+
+import akka.actor.Actor
+import akka.actor.actorRef2Scala
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.connector.MessageConsumer
+
+object ActivationFeed {
+    sealed class ActivationNotification
+
+    /** Pulls new messages from the message bus. */
+    case class FillQueueWithMessages()
+
+    /** Indicates resources are available because transaction completed, may cause pipeline fill. */
+    case class ContainerReleased(tid: TransactionId) extends ActivationNotification
+
+    /** Indicate resources are available because transaction failed, may cause pipeline fill. */
+    case class FailedActivation(tid: TransactionId) extends ActivationNotification
+}
+
+/**
+ * This actor polls the message bus for new messages and dispatches them to the given
+ * handler. The actor tracks the number of messages dispatched and will not dispatch new
+ * messages until some number of them are acknowledged.
+ *
+ * This is used by the invoker to pull messages from the message bus and apply back pressure
+ * when the invoker does not have resources to complete processing messages (i.e., no containers
+ * are available to run new actions).
+ *
+ * When the invoker releases resources (by reclaiming containers) it will send a message
+ * to this actor which will then attempt to fill the pipeline with new messages.
+ *
+ * The actor tries to fill the pipeline with additional messages while the number
+ * of outstanding requests is below the pipeline fill threshold.
+ */
+@throws[IllegalArgumentException]
+protected class ActivationFeed(
+    logging: Logging,
+    consumer: MessageConsumer,
+    maxPipelineDepth: Int,
+    longpollDuration: FiniteDuration,
+    handler: (String, Array[Byte]) => Any)
+    extends Actor {
+    import ActivationFeed.ActivationNotification
+    import ActivationFeed.FillQueueWithMessages
+
+    require(maxPipelineDepth >= 2 && maxPipelineDepth % 2 == 0, "max pipeline depth must be even and >= 2")
+    require(consumer.maxMessagePerPoll <= maxPipelineDepth, "consumer may yield more messages per poll than permitted by max depth")
+
+    private val pipelineFillThreshold = maxPipelineDepth - consumer.maxMessagePerPoll
+    private var pipelineOccupancy = 0
+
+    override def receive = {
+        case FillQueueWithMessages => Try {
+            val messages = consumer.getMessages(longpollDuration) // grab next batch of messages
+            val count = messages.size
+            messages foreach {
+                case (topic, partition, offset, bytes) =>
+                    logging.info(this, s"processing $topic[$partition][$offset ($count)]")(TransactionId.dispatcher)
+                    pipelineOccupancy += 1
+                    handler(topic, bytes)
+            }
+            fill()
```

Comment:
It would "work" with a thread pool of size 1 but not robustly if you mean "pool" generally.  If it's really a pool, that means someone else is part of the pool and they could get blocked for a long time on that call.